### PR TITLE
Add pure pursuit library

### PR DESCRIPTION
**Pull request by chfriedrich98**
### Solved Problem
The pure pursuit algorithm is used by the new ackermann rover module and soon also by the updated differential rover and the upcoming boat module.
Creating a library for the algorithm avoids having to copy the same code into all the modules.

### Solution
The algorithm calculates the intersection point between a circle around the vehicle and the line segment connecting the previous and current waypoint and returns the heading towards that point. The radius of the circle around the vehicle is used to tune the controller and is often referred to as look-ahead distance ($l_d$).
image

### Test coverage
- Unit test for edge cases
- SITL tested (ackermann rover module)

### CODE REVIEW

File: src/lib/pure_pursuit/PurePursuit.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+class PurePursuit
+{
+public:
+	/**
+	 * @brief Return heading towards the intersection point between a circle with a radius of
+	 * lookahead_distance around the vehicle and an extended line segment from the previous to the current waypoint.
+	 * Will fallback to returning the bearing to the current waypoint if inputs are invalid or if no
+	 * intersection points exist.
+	 * Exception: Will return 0 if curr_wp_local or curr_pos_local are invalid.
+	 * @param curr_wp_local Current waypoint in NED frame.
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.hpp:76]: why not `curr_wp_ned` then?

Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.hpp]: Makes sense, renamed in [f8d7531](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/f8d7531ed9098669f8c888ffdc455970978212c7).

File: src/lib/pure_pursuit/PurePursuit.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ ****************************************************************************/
+
+#include <matrix/math.hpp>
+
+using namespace matrix;
+
+/**
+ * @file PurePursuit.hpp
+ *
+ * Implementation of pure pursuit guidance logic.
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.hpp:41]: Could you add a reference to the formulation you are using? There are variants of the pure pursuit, and it would be nice if this is already clarified in the comments.

File: src/lib/pure_pursuit/PurePursuit.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+public:
+	/**
+	 * @brief Return heading towards the intersection point between a circle with a radius of
+	 * lookahead_distance around the vehicle and an extended line segment from the previous to the current waypoint.
+	 * Will fallback to returning the bearing to the current waypoint if inputs are invalid or if no
+	 * intersection points exist.
+	 * Exception: Will return 0 if curr_wp_ned or curr_pos_ned are invalid.
+	 * @param curr_wp_ned Current waypoint in NED frame.
+	 * @param prev_wp_ned Previous waypoint in NED frame.
+	 * @param curr_pos_ned Current position of the vehicle in NED frame.
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.hpp:78]: Units in meters?

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+#include "PurePursuit.hpp"
+#include <mathlib/mathlib.h>
+
+float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				      const Vector2f &curr_pos_ned,
+				      const float &lookahead_distance)
+{
+	// Check minimal necessary input validity
+	if (!PX4_ISFINITE(curr_wp_ned(0))
+	    || !PX4_ISFINITE(curr_wp_ned(1)) || !PX4_ISFINITE(curr_pos_ned(0)) || !PX4_ISFINITE(curr_pos_ned(1))) {
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:43]:
```suggestion
	if (curr_wp_ned.isAllFinite() && curr_pos_ned.isAllFinite()) {
```

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		return 0.f; // No fallback possible
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:44]: Does this mean that we cannot command a zero heading? Wouldn't make more sense to pass a NaN in case something is wrong?

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	// Setup fallback to bearing to the current waypoint
+	const Vector2f curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
+	const float bearing_to_curr_wp = atan2f(curr_pos_to_curr_wp(1), curr_pos_to_curr_wp(0));
+
+	// Check full input validity
+	if (lookahead_distance < FLT_EPSILON || !PX4_ISFINITE(lookahead_distance) || !PX4_ISFINITE(prev_wp_ned(0))
+	    || !PX4_ISFINITE(prev_wp_ned(1))) {
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:53]:
```suggestion
	if (lookahead_distance < FLT_EPSILON || !PX4_ISFINITE(lookahead_distance) || prev_wp_ned.isAllFinite()) {
```

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	// Catch waypoint overlap
+	if (fabsf(curr_wp_ned(0) - prev_wp_ned(0)) < FLT_EPSILON
+	    && fabsf(curr_wp_ned(1) - prev_wp_ned(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_ned(0) - prev_wp_ned(0) < FLT_EPSILON) {
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:64]: There are multiple instances of the same calculation `curr_wp_ned(0) - prev_wp_ned(0)`. We can probably do it once and use it repeatedly?

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		return atan2f(y0, x0);
+
+	} else { // Two intersetion points exist
+		const float d = lookahead_distance * lookahead_distance - line_segment_rover_offset * line_segment_rover_offset /
+				(line_segment_slope * line_segment_slope + 1.0f);
+		const float mult = sqrt(d / (line_segment_slope * line_segment_slope + 1.0f));
+		const Vector2f point1(x0 + mult, y0 + line_segment_slope * mult);
+		const Vector2f point2(x0 - mult, y0 - line_segment_slope * mult);
+		const Vector2f distance1 = (curr_wp_ned - curr_pos_ned) - point1;
+		const Vector2f distance2 = (curr_wp_ned - curr_pos_ned) - point2;
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:104]:
```suggestion
		const Vector2f distance1 = curr_pos_to_curr_wp - point1;
		const Vector2f distance2 = curr_pos_to_curr_wp - point2;
```

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]
Awesome!

Added some minor stylistic comments

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 5 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp:43]: Commanding a zero heading is not a problem.
This just catches the case where either the current waypoint or the current position are NaN (in this case I can't fallback to returning the bearing to the current waypoint).
Unfortunately the current position can become invalid at the times you wouldn't expect, for example it becomes NaN for a view update cycles after starting a mission.
I prefer returning zero here, because a NaN value can break things down the line. For example I had a slew rate in the ackermann module and if that is ever updated with a NaN value it permanently breaks.

File: src/lib/pure_pursuit/PurePursuit.hpp (hunk 2 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.hpp:41]: This library simply returns the heading to the intersection point between a circle around the vehicle and the line segment between the previous and next waypoint.
Correct me if I am wrong, but the differences between the pure pursuit variants are how this heading is translated into a steering input and how the lookahead radius adapts to i.e. different vehicle speeds. 
This is covered in the individual modules (since it is different for each) and not in the library.

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	// Catch waypoint overlap
+	if (fabsf(curr_wp_ned(0) - prev_wp_ned(0)) < FLT_EPSILON
+	    && fabsf(curr_wp_ned(1) - prev_wp_ned(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_ned(0) - prev_wp_ned(0) < FLT_EPSILON) {
+		if (fabsf(curr_wp_ned(0) - curr_pos_ned(0)) > lookahead_distance) { // No intersection point exists
+			return bearing_to_curr_wp; // Fallback
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.cpp:66]: why not going to the closest point on the line between prev and current wp?

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 7 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: Makes sense, implemented in [54e99ae](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/54e99ae35fcc2075d8e393ccf89084c19b1b509d).

File: src/lib/pure_pursuit/PurePursuit.hpp (hunk 3 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.hpp]: Yes, added clarification in [54e99ae](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/54e99ae35fcc2075d8e393ccf89084c19b1b509d).

**Comment by chfriedrich98** [comment]
@Jaeyoung-Lim Thanks for the review!
I implemented most of your suggestions and left some comments.

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+		} else {
+			const float x0 = curr_wp_ned(0) - curr_pos_ned(0);
+			const float y0 = sqrtf(powf(lookahead_distance, 2.f) - powf(x0, 2.f));
+
+			if (fabsf(curr_wp_ned(1) - (curr_pos_ned(1) + y0)) < fabsf(curr_wp_ned(1) - (curr_pos_ned(1) - y0))) {
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.cpp:72]: That's just to tell if we should go to the East or West intersection, right?
Then you could just say that 
```const float y0 = sign(curr_wp_ned(1) - curr_pos_ned(1)) * sqrtf(powf(lookahead_distance, 2.f) - powf(x0, 2.f));```

and only use `atan2f(y0, x0);`

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+	    && fabsf(curr_wp_ned(1) - prev_wp_ned(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_ned(0) - prev_wp_ned(0) < FLT_EPSILON) {
+		if (fabsf(curr_wp_ned(0) - curr_pos_ned(0)) > lookahead_distance) { // No intersection point exists
+			return bearing_to_curr_wp; // Fallback
+
+		} else {
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.cpp:68]: and what if the distance between the current wp and position is < lookahead_distance ? Don't you want to directly go to the current wp?

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/lib/pure_pursuit/PurePursuit.hpp (hunk 2 above)
Comment (Jaeyoung-Lim) [reply src/lib/pure_pursuit/PurePursuit.hpp:41]: I know, but if you can cite a paper that is consistent to the formulation you are using, it would be helpful for people understanding the implementation.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 5 above)
Comment (Jaeyoung-Lim) [reply src/lib/pure_pursuit/PurePursuit.cpp:43]: If you return a zero, there is no way outside of the library to detect whether if the input you have provided is valid or not.

It is actually a positive feature if something breaks down the line. It means that you need to handle these cases explicitly for your runtime application, rather than concealing it in the library.

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	const Vector2f curr_wp_to_prev_wp = curr_wp_ned - prev_wp_ned;
+
+	if (fabsf(curr_wp_to_prev_wp(0)) < FLT_EPSILON
+	    && fabsf(curr_wp_to_prev_wp(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_to_prev_wp(0) < FLT_EPSILON) {
+		if (fabsf(curr_wp_ned(0) - curr_pos_ned(0)) > lookahead_distance) { // No intersection point exists
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:65]:
```suggestion
		if (fabsf(curr_pos_to_curr_wp(0)) > lookahead_distance) { // No intersection point exists
```

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+			return bearing_to_curr_wp; // Fallback
+
+		} else {
+			const float x0 = curr_wp_ned(0) - curr_pos_ned(0);
```
Comment (Jaeyoung-Lim) [inline src/lib/pure_pursuit/PurePursuit.cpp:69]:
```suggestion
			const float x0 = curr_pos_to_curr_wp(0);
```

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 11 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp:68]: This is also an option, but has this tradeoff:
This will essentially reduce the lookahead radius until the vehicle enters the acceptance radius. If you have a small acceptance radius this can lead to oscillations (general problem with small lookahead).
Might still be the better option, I'll test it with the ackermann rover.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 9 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp:66]: Also a valid option. The most common scenario where this fallback happens is when the user manually changes the targeted waypoint during a mission which means the rover can be way off the line segment. 
I prefer the current fallback mainly because it reaches the current waypoint faster to continue the mission normally:
image
As a side note: In the ackermann module i scale the lookahead radius to the crosstrack error in this scenario (up to a maximum lookahead, after which I also just target the waypoint), which is equivalent to what you suggested.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 10 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: Yes, this is much cleaner thanks!
Implemented in 9e00187.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 5 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: That is a good point, implemented in 9e00187.

**Comment by chfriedrich98** [comment]
After discussing the implementation with @bresch I made some major changes to how the intersection point is calculated. It is now done using vectors which eliminates many of the edgecases I had to consider.
Also to keep everything consistent with

>If you return a zero, there is no way outside of the library to detect whether if the input you have provided is valid or not.
It is actually a positive feature if something breaks down the line. It means that you need to handle these cases explicitly for your runtime application, rather than concealing it in the library. The way it is implemented now, an invalid position estimate will result in aligning the vehicle into zero heading, which is a behavior that is unrelated to the failure.

the algorithm now returns NAN if any of the inputs are invalid rather than using the bearing to the current waypoint as a fallback.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 11 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: I now implemented this in 29b93d9, i agree that it is the better option.

File: src/lib/pure_pursuit/PurePursuit.cpp (hunk 9 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: After our discussion I changed this in 29b93d9 to your suggestion.

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+	const Vector2f prev_wp_to_curr_wp = curr_wp_ned - prev_wp_ned;
+
+	if (curr_pos_to_curr_wp.norm() < lookahead_distance
+	    || prev_wp_to_curr_wp.norm() <
+	    FLT_EPSILON) { // Target current waypoint if closer to it than lookahead or waypoints overlap
+		return atan2f(curr_pos_to_curr_wp(1), curr_pos_to_curr_wp(0));
+	}
+
+	const Vector2f prev_wp_to_curr_pos = curr_pos_ned - prev_wp_ned;
+	const Vector2f distance_on_line_segment = ((prev_wp_to_curr_pos * prev_wp_to_curr_wp) /
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.cpp:58]: you can use `Vector2f prev_wp_to_curr_wp_u = prev_wp_to_curr_wp.unit_or_zero()` to get a unit vector and then use it to compute the 
```distance_on_line_segment = (prev_wp_to_curr_pos * prev_wp_to_curr_wp_u) * prev_wp_to_curr_wp_u;```

and reuse it for
```prev_wp_to_intersection_point = distance_on_line_segment + line_extension * prev_wp_to_curr_wp_u;```

Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: implemented in [83bcedf](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/83bcedfc9056ab34bd00a244916fb7c74daf8e5f)

File: src/lib/pure_pursuit/PurePursuit.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "PurePursuit.hpp"
+#include <mathlib/mathlib.h>
+
+float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				      const Vector2f &curr_pos_ned,
+				      const float &lookahead_distance)
```
Comment (bresch) [inline src/lib/pure_pursuit/PurePursuit.cpp:39]: You shouldn't pas a native type by const ref, it's not more efficient, just pass by copy

Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.cpp]: changed in [ec4923f](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/ec4923f129fe29d3e21d639c281a2c3b36511420)

File: src/lib/pure_pursuit/PurePursuit.hpp (hunk 2 above)
Comment (chfriedrich98) [reply src/lib/pure_pursuit/PurePursuit.hpp:41]: added in [ec4923f](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/ec4923f129fe29d3e21d639c281a2c3b36511420)

**Review by bresch — APPROVED** [review]

**Review by Jaeyoung-Lim — APPROVED** [review]
Looks much nicer :)
